### ISSUE TITLE
bug on pinMode() and digitalWrite()

### ISSUE THREAD

**Issue by battosai30** [GrumpyOldPizza/ArduinoCore-stm32l0#52]
Hi, 

I hope you're fine :) I did not have the time to help you writing documentation or modifications but I hope to have time soon ... Upgrading my module I found a very strange bug.

If you remember I posted an issue months ago about timing issue of delayMicroseconds() which leads to fail using OneWire library. I saw you corrected this point and decided to upgrade my module. But I lost my DS18B20. I took my oscilloscope and I found some strange little spikes : 

![image](https://user-images.githubusercontent.com/5184702/49790202-f58fd580-fd2d-11e8-8749-441bc1093377.png)

As it should be more like : 

![image](https://user-images.githubusercontent.com/5184702/49790246-09d3d280-fd2e-11e8-9bab-03df3cde1a5e.png)

(it's not the same message and the same speed but there are both onewire communications).

So I tried to target the problem and found this : 

This code : 

```
void setup() {
}

void loop() {
 delayMicroseconds(10);
  pinMode(2,OUTPUT); // <=
 digitalWrite(2,LOW); // <=
 delayMicroseconds(10);
 digitalWrite(2,HIGH);
}
```

Leads to this : 

![image](https://user-images.githubusercontent.com/5184702/49790381-561f1280-fd2e-11e8-804d-bef17dc5c0f4.png)

This code : 

```
void setup() {
}

void loop() {
 delayMicroseconds(10);
 digitalWrite(2,LOW);  // <=
  pinMode(2,OUTPUT); // <=
 delayMicroseconds(10);
 digitalWrite(2,HIGH);
}
```

Leads to this : 

![image](https://user-images.githubusercontent.com/5184702/49790484-a1392580-fd2e-11e8-817f-8a106649cae6.png)

Context : 
-Board custom with Murata ABZ.
- I wrote my own variant but it's largelly like a grasshoper.
- As it's for onewire, there is a pull up resistor on pin 2. I tested on a pin without pull up resistor => seems ok, so it looks like a high impedance temporary status.

Regards

**Comment by battosai30** [comment]
Update : after hours and hours seeking, I found a wrokaround for OneWire : 
https://github.com/battosai30/OneWire_Workaround_0.0.9

Basically I made two type of modification : 
1) inverted order for pinMode() and digitalWrite() in write command (to avoid the issu I mentionned in my previous post).
2) Adjust timings : on this part, as I made a lot of tests, maybe not all timing needed to be changed but ... It works it works ! I think the strangest part was on the read_bit() which needs a really quick jump from OUTPUT to INPUT state, which seems to take a "long time" (but it's strange because looking  at signal with a digital analyzer showed enough time for reading vs DS18B20 specs but maybe it's quicker ...).

So for my first issue I think you will directly find the solution. For OneWire I think I will get some time to write the specific low level commands for direct driving, I just need to know if you created a #define somewhere in core for knowing we are compiling on your core (like #ifdef (__AVR__))

**Comment by GrumpyOldPizza** [comment]
I think this is more an issue at the system level code. There the logic is
to always go the INPUT before reconfiguring. Hence if you go from a OUTPUT
(with a LO value) to a INPUT with a pullup, then you trigger an edge. So
the code needs to be smarter and filter out redundant operations and such
ones that would cause unexpected edges.

**Comment by WizardAtLarge** [comment]
@battosai30 => I made some comments on your https://github.com/battosai30/OneWire_Workaround_0.0.9 repository that will make your code more robust. As it currently is written, the code introduces a bug in the expansion of "delay * TIMING_C" for values of delay < (10/1.2). Your code also violates minimum timing restrictions per Maxim's app notes

**Comment by GrumpyOldPizza** [comment]
Fixed in release 0.0.10.
